programs/drift_vaults/src/state/vault.rs

```diff
@@ -130,45 +148,209 @@ impl Vault {
+                }
+            }
+            Some(vp) => {
+                if self.management_fee != 0 && vp.protocol_fee != 0 && depositor_equity > 0 {
+                    let since_last = now.safe_sub(self.last_fee_update_ts)?;
+                    let total_fee = self
+                        .management_fee
+                        .safe_add(vp.protocol_fee.cast()?)?
+                        .cast::<i128>()?;
+
+                    // if protocol fee is non-zero and total fee would lead to zero equity remaining,
+                    // so tax equity - 1 but only for the protocol, so that the user is left with 1 and the manager retains their full fee.
+                    let total_fee_payment = depositor_equity
+                        .safe_mul(total_fee)?
+                        .safe_div(PERCENTAGE_PRECISION_I128)?
+                        .safe_mul(since_last.cast()?)?
+                        .safe_div(ONE_YEAR.cast()?)?;
+                    let management_fee_payment = total_fee_payment
+                        .safe_mul(self.management_fee.cast()?)?
+                        .safe_div(total_fee)?;
+                    let protocol_fee_payment = total_fee_payment
+                        .min(depositor_equity.saturating_sub(1))
+                        .safe_mul(vp.protocol_fee.cast()?)?
+                        .safe_div(total_fee)?;
+
+                    let new_total_shares_factor: u128 = depositor_equity
+                        .safe_mul(PERCENTAGE_PRECISION_I128)?
+                        .safe_div(
+                            depositor_equity
+                                .safe_sub(management_fee_payment)?
+                                .safe_sub(protocol_fee_payment)?,
+                        )?
+                        .cast()?;
+                    let mgmt_fee_shares_factor: u128 = depositor_equity
+                        .safe_mul(PERCENTAGE_PRECISION_I128)?
+                        .safe_div(depositor_equity.safe_sub(management_fee_payment)?)?
+                        .cast()?;
+                    let protocol_fee_shares_factor: u128 = depositor_equity
+                        .safe_mul(PERCENTAGE_PRECISION_I128)?
+                        .safe_div(depositor_equity.safe_sub(protocol_fee_payment)?)?
+                        .cast()?;
+
+                    let new_total_shares = self
+                        .total_shares
+                        .safe_mul(new_total_shares_factor.cast()?)?
+                        .safe_div(PERCENTAGE_PRECISION)?
+                        .max(self.user_shares);
+
+                    management_fee_shares = self
+                        .total_shares
+                        .safe_mul(mgmt_fee_shares_factor.cast()?)?
+                        .safe_div(PERCENTAGE_PRECISION)?
+                        .max(self.user_shares)
+                        .cast::<i128>()?
+                        .safe_sub(self.total_shares.cast()?)?;
+
+                    protocol_fee_shares = self
```

Comment:
think it might be better to do `new_total_shares - management_fee_shares - vault.total_shares` here so that we dont get weird numerical issues for rounding in multipication?

Reply:
this should stay. switching both of your comments to that math breaks the tests.
in the first comment that protocol calc does:
```rust
let protocol_fee_payment = total_fee_payment
  .min(depositor_equity.saturating_sub(1))
  .safe_sub(management_fee_payment)?;
```
This is because if the shares are odd, then the applied fee to the manager and protocol can't divide, so one of them must lose one shares. I make the protocol take the short end of the stick.

This second comment has to use the protocol_fee_shares_factor so that it has that subtraction factored in. It can't just do `total - manager = protocol` like you suggest.